### PR TITLE
Memory leak test for PHP unit tests

### PR DESCRIPTION
**Pull request by jerrycmh**
Add MemoryLeakTest.php in src/php/tests/MemoryLeakTest/MemoryLeakTest.php for running valgrind on those unit tests for memory leak report.

There are some reused class name between MemoryLeakTest.php and some of the unit test files. To avoid that, I create a MemoryLeakTest folder under tests directory to run it and unit tests separately. 

And also when running MemoryLeakTest.php together with those unit tests. There will raise a c-core error and one of the unit test will fail. Haven't looked through what is cause of the problem. But for now, I leave it separately with unit tests.

### CODE REVIEW

File: templates/tools/dockerfile/php7_deps.include [hunk 1]
```diff
@@ -19,6 +19,7 @@ RUN apt-get update && apt-get install -y ${'\\'}
   re2c ${'\\'}
   time ${'\\'}
   unzip ${'\\'}
+  valgrind ${'\\'}
```
Comment (jtattermusch) [inline templates/tools/dockerfile/php7_deps.include]: I'm not convinced valgrind should be part of PHP's standard testing dependencies (it then bubbles up e.g. into interop test images, where it is not needed). How big is valgrind installation?

Comment (stanley-cheung) [reply templates/tools/dockerfile/php7_deps.include]: @jtattermusch Thanks for the review.

So you are saying that, in addition to the `php` and `php7` base docker images we have now,  we should have a `php_memtest` base image, which should be completely separate from the rest, and only in there we install and run `valgrind`?

File: tools/dockerfile/test/multilang_jessie_x64/Dockerfile [hunk 2]
```diff
@@ -113,8 +113,11 @@ RUN /bin/bash -l -c "nvm alias default 10"
 # Install dependencies
 
 RUN apt-get update && apt-get install -y \
+	valgrind \
```
[the remark this answers is not in the document]
Comment (jerrycmh) [reply tools/dockerfile/test/multilang_jessie_x64/Dockerfile]: I just did a rebase on the freshest master

File: templates/tools/dockerfile/php7_deps.include (hunk 1 above)
Comment (jtattermusch) [reply templates/tools/dockerfile/php7_deps.include]: adding a new image is an overkill. But I'd like to see if we could skip installing valgrind where it is not necessary (e.g. for interop tests) and I'd like to know how much it adds to the PHP size.
Also please note that linux docker images are not the only place where we run PHP tests (e.g. there are performance workers and tests on other platforms) where valgrind is not installed - so using valgrind should probably be optional.

Comment (jerrycmh) [reply templates/tools/dockerfile/php7_deps.include]: valgrind adds 0.35GB to the image size.

Comment (stanley-cheung) [reply templates/tools/dockerfile/php7_deps.include]: So what we should do, is to add some logic to say, 1. if this is not interop test, 2. if it's linux only, install valgrind and run the memory leak test?

Comment (jtattermusch) [reply templates/tools/dockerfile/php7_deps.include]: Let's only add `<%include file="../../valgrind.include"/>` in the right docker images where they are useful. E.g. only add here after this line: https://github.com/grpc/grpc/blob/3278bdceda8030d5aa130f12765e5f07263c860d/templates/tools/dockerfile/test/php7_jessie_x64/Dockerfile.template#L21`

Comment (jerrycmh) [reply templates/tools/dockerfile/php7_deps.include]: @jtattermusch Thanks for helping. I just pushed changes into this PR followed by your comment and create a valgrind.include file for install valgrind command. But my question is what is the difference between adding include valgrind installation inside docker images and in php_deps.include ? Should <%include file="../../valgrind.include"/> will still always include valgrind installation?

File: templates/tools/dockerfile/valgrind.include [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#=================
```
Comment (jtattermusch) [inline templates/tools/dockerfile/valgrind.include]: nit: if the file contains PHP test dependencies, then it should be called sth like `php_valgrind.include` instead.

File: templates/tools/dockerfile/php7_deps.include (hunk 1 above)
Comment (jtattermusch) [reply templates/tools/dockerfile/php7_deps.include]: The difference is that like this valgrind won't be installed in the interop docker image (and interop docker images are being kept for backward compatibility suite).

File: src/php/bin/run_tests.sh [hunk 4]
```diff
@@ -22,9 +22,19 @@ cd src/php/bin
...
 
 php $extension_dir -d max_execution_time=300 $(which phpunit) -v --debug \
   ../tests/unit_tests/PersistentChannelTests
 
+export ZEND_DONT_UNLOAD_MODULES=1
+export USE_ZEND_ALLOC=0
+# Detect whether valgrind is executable
+if ! [ -x "$(command -v valgrind)" ]; then
+  echo 'Error: valgrind is not installed and is not executable' >&2
+  exit 1
```
Comment (stanley-cheung) [inline src/php/bin/run_tests.sh]: Let's do `exit 0`. Or, we can just move the main `valgrind` command below inside the `if` here, and flip the conditional. Like

```
if <valgrind is installed>
  run valgrind command
fi
```

File: src/php/tests/unit_tests/CallTest.php [hunk 5]
```diff
@@ -86,6 +86,26 @@ public function testAddSingleAndMultiValueMetadata()
...
+    {
+        $batch = [
+            Grpc\OP_SEND_INITIAL_METADATA => ['key1' => ['value1', 'value2'],
+                                              'key2' => ['value3', 'value4'],],
+        ];
+        $result = $this->call->startBatch($batch);
+        $this->assertTrue($result->send_metadata);
+    }
+
+    public function testAddMultiAndMultiValueMetadata()
```
Comment (stanley-cheung) [inline src/php/tests/unit_tests/CallTest.php]: Please remove one of these. They seem to be duplicate

File: src/php/bin/run_tests.sh [hunk 6]
```diff
@@ -22,9 +22,18 @@ cd src/php/bin
...
 php $extension_dir -d max_execution_time=300 $(which phpunit) -v --debug \
   ../tests/unit_tests/PersistentChannelTests
 
+export ZEND_DONT_UNLOAD_MODULES=1
+export USE_ZEND_ALLOC=0
+# Detect whether valgrind is executable
+if [ -x "$(command -v valgrind)" ]; then
+  valgrind --error-exitcode=10 --leak-check=yes php $extension_dir -d max_execution_time=300 \
+	../tests/MemoryLeakTest/MemoryLeakTest.php
+  exit 0
```
Comment (stanley-cheung) [inline src/php/bin/run_tests.sh]: One more nit: we don't need `exit 0` here now. We can let the script finish with whatever status the `valgrind` command finishes with.

File: src/php/tests/MemoryLeakTest/MemoryLeakTest.php [hunk 7]
```diff
@@ -0,0 +1,2310 @@
...
+        $this->channel = new \Grpc\Channel('localhost:50050', $opts);
+        return $this->channel;
+    }
+
+    public function UnaryCall($channel, $method, $deserialize, $options) {
+        return new UnaryCall($channel, $method, $deserialize, $options);
+    }
+
+    public function ClientStreamingCall($channel, $method, $deserialize, $options) {
+        return new ClientStreamingC
```
Comment (vjpai) [inline src/php/tests/MemoryLeakTest/MemoryLeakTest.php]: Why are there apparent version control markers in this file? This is blocking the import. We should probably have an OSS sanity check for this too, but this doesn't look right.

Comment (jerrycmh) [reply src/php/tests/MemoryLeakTest/MemoryLeakTest.php]: sorry, this is my bad.

**Comment by jtattermusch** [comment]
Another problem is that this was merged without checking test results.
One problem is that  the updated php docker images haven't been uploaded to dockerhub and so the PHP tests don't even start running:

https://source.cloud.google.com/results/invocations/095686f9-dadd-4cf5-8327-1138cf43ea33/targets/github%2Fgrpc%2Faggregate_tests/tests

```
+ DOCKER_IMAGE_NAME=php_jessie_x64_80146e3731996cce2cdd054cab60b0ead537a266
+ '[' grpctesting '!=' '' ']'
+ DOCKER_IMAGE_NAME=grpctesting/php_jessie_x64_80146e3731996cce2cdd054cab60b0ead537a266
+ docker pull grpctesting/php_jessie_x64_80146e3731996cce2cdd054cab60b0ead537a266
Using default tag: latest
Error response from daemon: pull access denied for grpctesting/php_jessie_x64_80146e3731996cce2cdd054cab60b0ead537a266, repository does not exist or may require 'docker login'
Traceback (most recent call last):
  File "tools/run_tests/run_tests.py", line 1575, in <module>
    env=env)
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'tools/run_tests/dockerize/build_docker_and_run_tests.sh' returned non-zero exit status 1
``` 

see https://github.com/grpc/grpc/blob/master/tools/dockerfile/push_testing_images.sh

**Comment by stanley-cheung** [comment]
Oh that's what the error is about. I spent some time today looking into this error without understanding it. Sorry about that.

**Pull request merged**